notebooks/02-ingestion/04-table-datatypes.ipynb

```diff
@@ -0,0 +1,851 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "6a116797-2dfb-4273-a50e-fda337bb3f7e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT\n",
+    "    COUNT(*) FILTER (WHERE \"DayofMonth_Filter\" = '10') AS \"10thDayCount\",\n",
+    "    COUNT(*) FILTER (WHERE \"DayofMonth_Filter\" = '9') AS \"9thDayCount\"\n",
+    "FROM \"example-flights-types-4\"\n",
+    "'''\n",
+    "\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "143b76e0-7c6a-4766-927c-4fb6ec82656f",
+   "metadata": {},
+   "source": [
+    "Building on the previous example, run the following cell to see a CAST using an internally stored string to compare with a numeric value:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "615fe582-ca47-4602-a1a6-383a0c4ca2bb",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT\n",
+    "    COUNT(*) FILTER (WHERE \"DayofMonth_Filter\" = '10') AS \"10thDayCount\",\n",
+    "    COUNT(*) FILTER (WHERE \"DayofMonth_Filter\" = '9') AS \"9thDayCount-1\",\n",
+    "    COUNT(*) FILTER (WHERE CAST(\"DayofMonth_Filter\" AS BIGINT) = 9) AS \"9thDayCount-2\",\n",
+    "    COUNT(*) FILTER (WHERE CAST(\"DayofMonth_Filter\" AS BIGINT) = 8) AS \"8thDayCount\"\n",
+    "FROM \"example-flights-types-4\"\n",
+    "'''\n",
+    "\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4f42e716-592a-42aa-bfa2-d743f963983f",
+   "metadata": {},
+   "source": [
+    "You can also convert from a TIMESTAMP to a string.\n",
+    "\n",
+    "Run the following cell to see a CAST from TIMESTAMP to a string, as well as an example of the TIME_FORMAT function to create the same output.\n",
+    "\n",
+    "> For more examples of TIME_FORMAT check out the dedicated notebook on [date and time functions](../03-query/07-functions-datetime.ipynb)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f2cbad55-b547-42e1-8713-6e1f6144c1b2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT\n",
+    "    CAST(\"__time\" AS VARCHAR) AS \"__time_VARCHAR\",\n",
+    "    TIME_FORMAT(__time, 'YYYY-MM-dd hh:mm:ss') AS \"__time_FORMAT\"\n",
+    "FROM \"example-flights-types-4\"\n",
+    "WHERE TIME_IN_INTERVAL(\"__time\",'2005-11-08/PT10M')\n",
+    "'''\n",
+    "\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a112ea9e-8f83-4adb-869a-498051b854e4",
+   "metadata": {},
+   "source": [
+    "## Arrays\n",
+    "\n",
+    "[Arrays](https://druid.apache.org/docs/latest/querying/sql-data-types#arrays) are specially recognised in Druid, constructed using appropriate [aggregations](https://druid.apache.org/docs/latest/querying/sql-aggregations) and then worked with using dedicated SQL [functions](https://druid.apache.org/docs/latest/querying/sql-array-functions).\n",
```

Comment:
in theory, you can also ingest them directly now either using "auto" data type in native or an ARRAY expression in MSQ.... I haven't tested this, but this can be R28 stuff.

Reply:
I'll update this notebook with a link out to the arrays notebook when I get it in for 28.